### PR TITLE
General/codequality

### PR DESCRIPTION
**Pull request by mwever**
* added GGP as hpo algo
* general code quality improvements
* extended statisticsreport to output prediction time of finally selected classifier in ML-Plan CLI
* search space fixes
* search space now includes list interfaces for the configuration of heterogeneous meta learners

### CODE REVIEW

File: JAICore/jaicore-algorithminspector/src/main/java/ai/libs/jaicore/graphvisualizer/plugin/graphview/GraphViewPluginModel.java [hunk 1]
```diff
@@ -72,6 +72,7 @@ private void initializeGraph(final File searchGraphCSSPath) {
 		this.graph = new SingleGraph("Search Graph");
 
 		try {
+			System.out.println(searchGraphCSSPath.getPath());
```
Comment (tornede) [inline JAICore/jaicore-algorithminspector/src/main/java/ai/libs/jaicore/graphvisualizer/plugin/graphview/GraphViewPluginModel.java:75]: syso

File: JAICore/jaicore-algorithminspector/src/main/java/ai/libs/jaicore/graphvisualizer/events/recorder/AlgorithmEventHistoryEntry.java [hunk 2]
```diff
@@ -1,89 +1,92 @@
...
+/**
+ * {@link AlgorithmEventHistoryEntry}s are used to store {@link IPropertyProcessedAlgorithmEvent}s in an {@link AlgorithmEventHistory} combined with additional meta information.
+ *
+ * @author atornede
+ *
+ */
+public class AlgorithmEventHistoryEntry implements Serializable {
+
+	/**
+	 *
```
Comment (alexandertornede) [inline JAICore/jaicore-algorithminspector/src/main/java/ai/libs/jaicore/graphvisualizer/events/recorder/AlgorithmEventHistoryEntry.java:16]: random comment

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+ * HyperBand is a simple but effective hyper-parameter optimization technique, heavily relying on a technique called successive halving.
+ * Given a maximum amount of allocatable resources r_max and an integer parameter eta > 1, it allocates resources in a clever way, racing
+ * randomly sampled solution candidates with increasing resources for more promising ones.
+ *
+ * For more details, refer to the published paper by Li et al. from 2018:
+ * Hyperband: A Novel Bandit-Based Approach to Hyperparameter Optimization. In: Journal of Machine Learning research 18 (2018) 1-52
+ *
+ * @author mwever
+ *
+ */
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:54]: CP comment

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+					} finally {
+						finished.release();
+					}
+				}
+			});
+			evoThread.start();
+
+			try {
+				if (this.getConfig().getTimeout().milliseconds() > 0) { // a timeout is specified, thus, finish with what occurs first: timeout or max generations reached
+					System.out.println("Wait for " + this.getConfig().getTimeout().milliseconds() + "ms");
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:220]: Syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+					boolean acquired = finished.tryAcquire(this.getConfig().getTimeout().milliseconds(), TimeUnit.MILLISECONDS);
+					if (!acquired) {
+						System.err.println("Timeout occurred for evo thread. Now interrupt it.");
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:223]: Syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+						evoThread.interrupt();
+					}
+				} else { // no timeout configured: wait until max generations are reached.
+					finished.acquire();
+				}
+			} catch (InterruptedException e) {
+				System.err.println("Main GGP thread got interrupted, now interrupt evoThread.");
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:230]: Syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+									GrammarBasedGeneticProgramming.this.earlyStoppingCounter.set(0);
+								}
+								realInd.setFitnessValue(fitnessValue);
+							} catch (ObjectEvaluationFailedException | InterruptedException e) {
+								realInd.setFitnessValue(10000.0);
+							} finally {
+								semaphore.release();
+							}
+						} catch (Throwable e) {
+							System.err.println(individual.toString());
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:288]: Syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+import ai.libs.jaicore.basic.IOwnerBasedAlgorithmConfig;
+import ai.libs.jaicore.basic.StatisticsUtil;
+import ai.libs.jaicore.basic.algorithm.AOptimizer;
+import ai.libs.jaicore.components.api.IComponentInstance;
+import ai.libs.jaicore.components.model.ComponentInstance;
+import ai.libs.jaicore.components.model.SoftwareConfigurationProblem;
+import ai.libs.jaicore.ml.hpo.ggp.GrammarBasedGeneticProgramming.GGPSolutionCandidate;
+
+/**
+ * HyperBand is a simple but effective hyper-parameter optimization technique, heavily relying on a technique called successive halving.
```
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:45]: HyperBand ?

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+		} catch (InterruptedException e) {
+			interrupted.set(true);
+			pool.shutdownNow();
+			throw e;
+		}
+		population.stream().forEach(x -> {
+			try {
+				this.cacheMap.put(x.toString(), x.getFitness());
+			} catch (Exception e) {
+				// could not cache fitness => nullpointer exception must have occurred, so ignore the candidate and do not put it into the cache.
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:308]: maybe log something here?

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+								}
+							}
+
+							List<CandidateProgram> offspring = new ArrayList<>(GrammarBasedGeneticProgramming.this.getConfig().getPopulationSize());
+							// keep elite
+							for (int i = 0; i < GrammarBasedGeneticProgramming.this.getConfig().getElitismSize(); i++) {
+								offspring.add(population.get(i));
+							}
+							if (Thread.interrupted()) {
+								System.out.println("Thread got interrupted, cancel GGP.");
```
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:163]: syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+									if (offspring.size() < GrammarBasedGeneticProgramming.this.getConfig().getPopulationSize()) {
+										offspring.add(child2);
+									}
+								}
+							}
+							population = offspring;
+							g++;
+						}
+					} catch (InterruptedException e) {
+						System.err.println("GGP thread got interrupted, release semaphore and shutdown.");
```
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:208]: syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java (hunk 4 above)
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:220]: syso

File: JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgrammingTest.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+		};
+
+		input = new SoftwareConfigurationProblem<>(components, requiredInterface, evaluator);
+	}
+
+	@Test
+	public void randomEvalTest() throws AlgorithmTimeoutedException, InterruptedException, AlgorithmExecutionCanceledException, AlgorithmException {
+		GrammarBasedGeneticProgramming ggp = new GrammarBasedGeneticProgramming(input);
+		GGPSolutionCandidate sol = ggp.call();
+		System.out.println(sol.getScore() + " " + sol.getComponentInstance());
```
Comment (alexandertornede) [inline JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgrammingTest.java:49]: syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java (hunk 5 above)
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:223]: syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java (hunk 6 above)
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:230]: syso

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java (hunk 7 above)
Comment (tornede) [inline JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:288]: syso

File: JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/classification/loss/instance/LogLossTest.java [hunk 13]
```diff
@@ -19,4 +21,12 @@ public void testLogLoss() {
 		assertEquals("Log loss does not work as expected.", -Math.log(1.0 - 1E-15), ll.loss(2, PREDICTED), 1E-8);
 	}
 
+	@Test
+	public void test() {
+		double[] exp = { 0.0, 1.0, 0.0 };
+		double[] pred = { 0.5, 0.999999, 0.5 };
+		System.out.println(Math.log(1));
```
Comment (tornede) [inline JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/classification/loss/instance/LogLossTest.java:28]: syso

File: softwareconfiguration/mlplan/full/experiment-setup.properties [hunk 14]
```diff
@@ -0,0 +1,23 @@
+mem.max=30720
+mem.opp=30720
+cpu.max=8
+
+db.driver=mysql
+db.host=localhost
+db.username=results
+db.password=Hallo333!
```
Comment (alexandertornede) [inline softwareconfiguration/mlplan/full/experiment-setup.properties:8]: DB

File: JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgrammingTest.java (hunk 12 above)
Comment (tornede) [inline JAICore/jaicore-ml/src/test/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgrammingTest.java:49]: syso

File: softwareconfiguration/mlplan/full/testrsc/report/statistics-report/regression.json [hunk 15]
```diff
@@ -0,0 +1,8 @@
+{
+  "num_evaluations" : 0,
```
Comment (alexandertornede) [inline softwareconfiguration/mlplan/full/testrsc/report/statistics-report/regression.json:2]: Are all of these statistics file supposed to be in here?

File: softwareconfiguration/mlplan/full/experiment-setup.properties (hunk 14 above)
Comment (tornede) [inline softwareconfiguration/mlplan/full/experiment-setup.properties]: public password

**Review by alexandertornede — CHANGES_REQUESTED** [review]
See comments

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+				.collect(Collectors.toList());
+		return argumentsList.stream();
+	}
+
+	@Disabled
+	@MediumParameterizedTest
+	@MethodSource("getComponentsToTestRandomConfigurations")
+	public void testRandomConfigurationsInstantiation(final String componentName, final int numParameterizations) throws IOException, ComponentInstantiationFailedException {
+		IComponent compToTest = repo.getComponent(componentName);
+		System.out.println("Test " + numParameterizations + " random configuration of component: " + compToTest.getName());
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:74]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+				IWekaClassifier learner = factory.getComponentInstantiation(ci);
+				ILearnerRunReport report = new SupervisedLearnerExecutor().execute(learner, regressionDataset, regressionDataset);
+			} catch (Exception e) {
+				System.out.println(ComponentInstanceUtil.getComponentInstanceString(ci));
+				e.printStackTrace();
+				fail(e);
+			}
+
+			if ((i + 1) % 10 == 0) {
+				System.out.println((i + 1) + " components tested");
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:96]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+	public static Stream<Arguments> getComponentsToTestRandomInstantiations() {
+		return ComponentUtil.getComponentsProvidingInterface(repo, "MetaRegressor").stream().map(x -> Arguments.of(x.getName(), NUM_PARAMETRIZATIONS));
+	}
+
+	@MediumParameterizedTest
+	@MethodSource("getComponentsToTestRandomInstantiations")
+	public void testRandomInstantiation(final String componentName, final int numInstantiations) throws JsonProcessingException {
+		IComponent compToTest = repo.getComponent(componentName);
+		System.out.println("Test " + numInstantiations + " random instantiation of component: " + compToTest);
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:109]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+			if (componentName.toLowerCase().contains("randomforest")) {
+				if (Integer.parseInt(ci.getParameterValues().get("N")) > regressionDataset.size()) {
+					continue;
+				}
+			}
+			try {
+				IWekaClassifier learner = factory.getComponentInstantiation(ci);
+				ILearnerRunReport report = new SupervisedLearnerExecutor().execute(learner, regressionDataset, regressionDataset);
+			} catch (Exception e) {
+				System.out.println("Failed for: " + ComponentInstanceUtil.getComponentInstanceString(ci));
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:125]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+				e.printStackTrace();
+				fail(e);
+			}
+
+			if ((i + 1) % 10 == 0) {
+				System.out.println((i + 1) + " components tested");
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:131]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/TestRegressionAlgos.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+
+		List<ILabeledDataset<ILabeledInstance>> dataset = OpenMLDatasetReader.loadTaskFold(3005, 0);
+		IComponentRepository repo = new ComponentSerialization().deserializeRepository(new ResourceFile("automl/searchmodels/weka/weka-full-regression.json"));
+		Collection<IComponentInstance> asChoices = ComponentUtil.getAllAlgorithmSelectionInstances("MetaRegressor", repo);
+
+		WekaRegressorFactory factory = new WekaRegressorFactory();
+		for (IComponentInstance ci : asChoices) {
+			try {
+				IWekaClassifier learner = factory.getComponentInstantiation(ci);
+				System.out.println("Test " + ComponentInstanceUtil.getComponentInstanceAsComponentNames(ci));
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/TestRegressionAlgos.java:35]: syso

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/TestRegressionAlgos.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+				learner.fit(dataset.get(0));
+				ILearnerRunReport report = new SupervisedLearnerExecutor().execute(learner, dataset.get(1));
+				System.out.println("RMSE: " + ERegressionPerformanceMeasure.RMSE.loss(report.getPredictionDiffList().getCastedView(Double.class, IRegressionPrediction.class)));
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/TestRegressionAlgos.java:38]: syso

File: softwareconfiguration/mlplan/full/src/main/java/ai/libs/mlplan/cli/MLPlanCLI.java [hunk 23]
```diff
@@ -203,6 +229,15 @@ private static void runMLPlan(final CommandLine cl) throws Exception {
 			System.err.println("Cannot use both: local dataset and openml task. Only one option either " + O_FIT_DATASET + " or " + O_OPENML_TASK + " may be given.");
```
Comment (tornede) [inline softwareconfiguration/mlplan/full/src/main/java/ai/libs/mlplan/cli/MLPlanCLI.java:229]: syso ;)

File: softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+			if (componentName.toLowerCase().contains("randomforest")) {
+				if (Integer.parseInt(ci.getParameterValues().get("N")) > 8) {
+					continue;
+				}
+			}
+			try {
+				IWekaClassifier learner = factory.getComponentInstantiation(ci);
+				ILearnerRunReport report = new SupervisedLearnerExecutor().execute(learner, regressionDataset, regressionDataset);
+			} catch (Exception e) {
+				System.out.println(ComponentInstanceUtil.getComponentInstanceString(ci));
```
Comment (tornede) [inline softwareconfiguration/mlplan/weka/src/test/java/ai/libs/mlplan/wekamlplan/searchspace/RandomComponentValidityTest.java:90]: syso

**Review by tornede — CHANGES_REQUESTED** [review]

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java (hunk 9 above)
Comment (mwever) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/hpo/ggp/GrammarBasedGeneticProgramming.java:308]: No if the evaluation failed, there has already been a log output in the evaluation block. So at this point it is only about caching individual's fitness values for convenience (to not re-evaluate the same individuals over and over again).

File: softwareconfiguration/mlplan/full/testrsc/report/statistics-report/regression.json (hunk 15 above)
Comment (mwever) [reply softwareconfiguration/mlplan/full/testrsc/report/statistics-report/regression.json:2]: yes, they are used to assert that the statistics reports are correctly produced from the results obtained by ml-plan.

File: softwareconfiguration/mlplan/full/src/main/java/ai/libs/mlplan/cli/MLPlanCLI.java (hunk 23 above)
Comment (mwever) [reply softwareconfiguration/mlplan/full/src/main/java/ai/libs/mlplan/cli/MLPlanCLI.java:229]: this is the CLI of ml-plan. I'd argue that for CLI this is better than a logger.

**Review by tornede — APPROVED** [review]

**Review by alexandertornede — APPROVED** [review]

**Pull request merged**